src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivitiesResourceIntTest.kt

```diff
@@ -882,25 +904,195 @@ class ActivitiesResourceIntTest : IntegrationTestBase() {
       }
     }
+  }
+
+  @Nested
+  inner class EndOffenderProgramProfile {
+
+    lateinit var courseActivity: CourseActivity
+    var bookingId: Long = 0
+
+    private val endOffenderProgramProfileRequest: () -> EndOffenderProgramProfileRequest = {
+      EndOffenderProgramProfileRequest(
+        endDate = LocalDate.parse("2023-01-28"),
+        endReason = "REL",
+        endComment = "A comment",
+      )
+    }
+
+    @BeforeEach
+    internal fun setup() {
+      courseActivity = repository.save(courseActivityBuilderFactory.builder(prisonId = prisonId))
+      bookingId = offenderAtMoorlands.latestBooking().bookingId
+      callCreateEndpoint(courseActivity.courseActivityId, bookingId)
+    }
+
+    @Test
+    fun `access forbidden when no authority`() {
+      webTestClient.put().uri("/activities/${courseActivity.courseActivityId}/booking-id/$bookingId/end")
+        .body(BodyInserters.fromValue(endOffenderProgramProfileRequest()))
+        .exchange()
+        .expectStatus().isUnauthorized
+    }
+
+    @Test
+    fun `access forbidden when no role`() {
+      webTestClient.put().uri("/activities/${courseActivity.courseActivityId}/booking-id/$bookingId/end")
+        .headers(setAuthorisation(roles = listOf()))
+        .body(BodyInserters.fromValue(endOffenderProgramProfileRequest()))
+        .exchange()
+        .expectStatus().isForbidden
+    }
 
-    private fun callCreateEndpoint(courseActivityId: Long, bookingId: Long): Long {
-      val response = webTestClient.post().uri("/activities/$courseActivityId")
+    @Test
+    fun `access forbidden with wrong role`() {
+      webTestClient.put().uri("/activities/${courseActivity.courseActivityId}/booking-id/$bookingId/end")
+        .headers(setAuthorisation(roles = listOf("ROLE_BANANAS")))
+        .body(BodyInserters.fromValue(endOffenderProgramProfileRequest()))
+        .exchange()
+        .expectStatus().isForbidden
+    }
+
+    @Test
+    fun `course activity not found`() {
+      val response = webTestClient.put().uri("/activities/999888/booking-id/$bookingId/end")
         .headers(setAuthorisation(roles = listOf("ROLE_NOMIS_ACTIVITIES")))
-        .contentType(MediaType.APPLICATION_JSON)
-        .body(
-          BodyInserters.fromValue(
-            """{
-            "bookingId"  : "$bookingId",
-            "startDate"  : "2022-11-14",
-            "endDate"    : "${TEN_DAYS_TIME.format(DateTimeFormatter.ISO_DATE)}",
-            "payBandCode": "5"
-          }"""
-          )
-        )
+        .body(BodyInserters.fromValue(endOffenderProgramProfileRequest()))
         .exchange()
-        .expectStatus().isCreated
-        .expectBody(CreateOffenderProgramProfileResponse::class.java)
+        .expectStatus().isNotFound
+        .expectBody(ErrorResponse::class.java)
+        .returnResult().responseBody
+      assertThat(response?.userMessage).isEqualTo("Not Found: Course activity with id=999888 does not exist")
```

Comment:
if you want you could use jsonPath natively to wettest client, no need to return a result

Reply:
ah yes, slightly simpler